File: 2016/02_bathroom_security/aoc02.py
"""Bathroom Security

Advent of Code 2016, day 2
Solution by Geir Arne Hjelle, 2016-12-03
"""
import pathlib
import sys

# Keypad for first part:     1 2 3
#                            4 5 6
#                            7 8 9
#
KEYPAD_1 = {
    1: {"D": 4, "R": 2},
    2: {"D": 5, "R": 3, "L": 1},
    3: {"D": 6, "L": 2},
    4: {"U": 1, "D": 7, "R": 5},
    5: {"U": 2, "D": 8, "R": 6, "L": 4},
    6: {"U": 3, "D": 9, "L": 5},
    7: {"U": 4, "R": 8},
    8: {"U": 5, "R": 9, "L": 7},
    9: {"U": 6, "L": 8},
}

# Keypad for second part:           1
#                                 2 3 4
#                               5 6 7 8 9
#                                 A B C
#                                   D
#
KEYPAD_2 = {
    1: {"D": 3},
    2: {"D": 6, "R": 3},
    3: {"U": 1, "D": 7, "R": 4, "L": 2},
    4: {"D": 8, "L": 3},
    5: {"R": 6},
    6: {"U": 2, "D": "A", "R": 7, "L": 5},
    7: {"U": 3, "D": "B", "R": 8, "L": 6},
    8: {"U": 4, "D": "C", "R": 9, "L": 7},
    9: {"L": 8},
    "A": {"U": 6, "R": "B"},
    "B": {"U": 7, "D": "D", "R": "C", "L": "A"},
    "C": {"U": 8, "L": "B"},
    "D": {"U": "B"},
}


def find_code(fid, keypad):
    key = 5
    code = []
    for line in fid:
        for instruction in line.strip():
            key = keypad[key].get(instruction, key)
        code.append(str(key))
    return "".join(code)


def main(args):
    """Solve the problem for all file paths"""
    for file_path in [pathlib.Path(p) for p in args if not p.startswith("-")]:
        solve(file_path)


def solve(file_path):
    """Solve the problem for one file path"""
    print(f"\n{file_path}:")
    with file_path.open(mode="r") as fid:
        print(f"Keypad 1: {find_code(fid, KEYPAD_1)}")

    with file_path.open(mode="r") as fid:
        print(f"Keypad 2: {find_code(fid, KEYPAD_2)}")


if __name__ == "__main__":
    main(sys.argv[1:])

File: old/2020/02_password_philosophy/aoc02.jl
# Password Philosophy
#
# Advent of Code 2020, day 2
# Solution by Geir Arne Hjelle, 2020-12-02

module AOC02

using Pipe


# Parse each line into parts: "{n1}-{n2} {char}: {word}"
function parse_line(line)
    pattern = r"^(?P<n1>\d+)-(?P<n2>\d+) (?P<char>.): (?P<word>.+)$"
    parts = @pipe line |> match(pattern, _)
    Dict(
        "num_1" => parse(Int16, parts[:n1]),
        "num_2" => parse(Int16, parts[:n2]),
        "char" => parts[:char][1],
        "word" => parts[:word],
    )
end


# Check password based on first policy: the given character must occur between
# num_1 and num_2 times.
function is_valid_1(parts)
    char, word = parts["char"], parts["word"]
    parts["num_1"] <= count(c -> (c == char), word) <= parts["num_2"]
end


# Check password based on second policy: Exactly one of characters at indices
# num_1 and num_2 must match the given character.
function is_valid_2(parts)
    char, word = parts["char"], parts["word"]
    (word[parts["num_1"]] == char) ⊻ (word[parts["num_2"]] == char)
end


# Solve the problem for one file
function solve(input)
    # Read from file
    passwords = input |> i -> split(i, "\n") .|> parse_line

    # Part 1
    part_1 = passwords .|> is_valid_1 |> sum

    # Part 2
    part_2 = passwords .|> is_valid_2 |> sum

    part_1, part_2
end


# Solve the problem for one file
function solve_file(file_path)
    println("\n$(file_path)")
    input = open(file_path) do fid
        read(fid, String) |> strip
    end
    input .|> solve
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module
